src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -33,6 +33,11 @@ namespace Microsoft.CodeAnalysis
     public abstract partial class Compilation
     {
         /// <summary>
+        /// For use only in <see cref="Microsoft.CodeAnalysis.Collections.CompilationStuff"/>.
+        /// </summary>
+        internal ConcurrentDictionary<object, object> _compilationStuff = new ConcurrentDictionary<object, object>();
```

Comment:
This could be lazily initialized with LazyInitializer.EnsureInitialized so you don't pay an allocation penalty unless it's used.